readme.md
# doctype [![Build Status](https://img.shields.io/travis/wooorm/doctype.svg?style=flat)](https://travis-ci.org/wooorm/doctype) [![Coverage Status](https://img.shields.io/codecov/c/github/wooorm/doctype.svg)](https://codecov.io/github/wooorm/doctype)

Information for HTML doctypes.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
npm install doctype
```

**doctype** is also available for
[bower](http://bower.io/#install-packages), [duo](http://duojs.org/#getting-started),
and for AMD, CommonJS, and globals ([uncompressed](doctype.js) and
[compressed](doctype.min.js)).

## Usage

```js
console.log(doctype(5));
// 'html'

console.log(doctype(4.01));
console.log(doctype('4.01t'));
console.log(doctype('4.01 Transitional'));
console.log(doctype('HTML 4.01 Transitional'));
// 'HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd"'

console.log(doctype('svg'));
// 'svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"'
```

## API

### doctype(name)

Get a doctype.

Parameters:

*   `name` (`string`) — Case-insensitive, white-space insensitive name:
    stripped from white-space, casing, fractional-part of version, `'HTML'`
    or `'XHTML'` prefix, and suffixes are normalised (`Transitional` to
    `t`).

Returns: `string?` — When applicable.

### doctype.all

```js
{
  'HTML 5': 'html',
  'HTML 4.01 Strict': 'HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd"',
  // ...
}
```

`Object` mapping doctype names to doctype strings. This gives raw access to
the information returned by [`doctype()`](#doctypename).

## Support

| Name                   | Doctype                                                                                                            |
| ---------------------- | ------------------------------------------------------------------------------------------------------------------ |
| HTML 5                 | `html`                                                                                                             |
| HTML 4.01 Strict       | `HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "<http://www.w3.org/TR/html4/strict.dtd>"`                                |
| HTML 4.01 Transitional | `HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "<http://www.w3.org/TR/html4/loose.dtd>"`                    |
| HTML 4.01 Frameset     | `HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN" "<http://www.w3.org/TR/html4/frameset.dtd>"`                     |
| HTML 3.2               | `HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN"`                                                                     |
| HTML 3.2               | `HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN"`                                                                     |
| HTML 2.0               | `HTML PUBLIC "-//IETF//DTD HTML//EN"`                                                                              |
| XHTML 1.1              | `html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "<http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd>"`                         |
| XHTML 1.0 Strict       | `html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "<http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd>"`             |
| XHTML 1.0 Transitional | `html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "<http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd>"` |
| XHTML 1.0 Frameset     | `html PUBLIC "-//W3C//DTD XHTML 1.0 Frameset//EN" "<http://www.w3.org/TR/xhtml1/DTD/xhtml1-frameset.dtd>"`         |
| XHTML 1.1              | `html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd"`                           |
| XHTML Basic 1.1        | `html PUBLIC "-//W3C//DTD XHTML Basic 1.1//EN" "http://www.w3.org/TR/xhtml-basic/xhtml-basic11.dtd"`               |
| XHTML Basic 1.0        | `html PUBLIC "-//W3C//DTD XHTML Basic 1.0//EN" "http://www.w3.org/TR/xhtml-basic/xhtml-basic10.dtd"`               |
| MathML 2.0             | `math PUBLIC "-//W3C//DTD MathML 2.0//EN" "http://www.w3.org/Math/DTD/mathml2/mathml2.dtd"`                        |
| MathML 1.01            | `math SYSTEM "http://www.w3.org/Math/DTD/mathml1/mathml.dtd"`                                                      |
| SVG 1.0                | `svg PUBLIC "-//W3C//DTD SVG 1.0//EN" "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd"`                  |
| SVG 1.1 Full           | `svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"`                          |
| SVG 1.1 Basic          | `svg PUBLIC "-//W3C//DTD SVG 1.1 Basic//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-basic.dtd"`              |
| SVG 1.1 Tiny           | `svg PUBLIC "-//W3C//DTD SVG 1.1 Tiny//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-tiny.dtd"`                |

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


history.md
<!--mdast setext-->

<!--lint disable no-multiple-toplevel-headings -->

1.1.0 / 2015-10-10
==================

*   Add more doctypes, including MathML, SVG ([3b1970c](https://github.com/wooorm/doctype/commit/3b1970c))

1.0.1 / 2015-10-09
==================

*   Fix type ([5355688](https://github.com/wooorm/doctype/commit/5355688))
*   Fix `doctype.doctypes` > `doctype.all` ([2ae9afb](https://github.com/wooorm/doctype/commit/2ae9afb))
*   Fix missing json ([71a3ccf](https://github.com/wooorm/doctype/commit/71a3ccf))

1.0.0 / 2015-10-09
==================


test.js
/**
 * @author Titus Wormer
 * @copyright 2015 Titus Wormer
 * @license MIT
 * @module doctype
 * @fileoverview Test suite for `doctype`.
 */

'use strict';

/* eslint-env node */

/*
 * Dependencies.
 */

var test = require('tape');
var doctype = require('./');

/*
 * Tests.
 */

test('doctype(name)', function (t) {
    t.deepEqual(doctype('foo'), null);

    t.deepEqual(doctype(5), 'html');
    t.deepEqual(doctype('5'), 'html');
    t.deepEqual(doctype('5.0'), 'html');
    t.deepEqual(doctype('HTML5'), 'html');
    t.deepEqual(doctype('HTML 5'), 'html');
    t.deepEqual(doctype('html 5'), 'html');
    t.deepEqual(doctype('html'), 'html');

    t.deepEqual(doctype(4), 'HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"');
    t.deepEqual(doctype(4.01), 'HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"');
    t.deepEqual(doctype('4.01t'), 'HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"');
    t.deepEqual(doctype('4.01 transitional'), 'HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"');
    t.deepEqual(doctype('4.01Transitional'), 'HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"');

    t.deepEqual(doctype('svg'), 'svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"');

    t.deepEqual(doctype('m'), 'math PUBLIC "-//W3C//DTD MathML 2.0//EN" "http://www.w3.org/Math/DTD/mathml2/mathml2.dtd"');

    t.deepEqual(doctype.all, require('./index.json'));

    t.end();
});


index.js
/**
 * @author Titus Wormer
 * @copyright 2015 Titus Wormer
 * @license MIT
 * @module doctype
 * @fileoverview Information for HTML doctypes.
 */

'use strict';

/* eslint-env commonjs */

/*
 * Dependencies.
 */

var doctypes = require('./index.json');

/** Get the first character of the second parameter. */
function character(_, value) {
    return value.charAt(0);
}

/** Clean and simplify a doctype name. */
function cleanDoctype(name) {
    return String(name)
        .toLowerCase()
        .replace(/([a-z]+|\d+)/, ' $1 ')
        .replace(/\.0+(?!\d)/, '')
        .replace(/\.01/, '')
        .replace(/\.2/, '')
        .replace(/\b(xhtml|html|mathml|svg|full|basic|tiny|strict|frameset|basic)\b/g, character)
        .replace(/ t(?:ransitional)?/, '')
        .replace(/\s+/g, '');
}

/*
 * Gather all doctypes as shortcodes.
 */

var shortcodes = {};
var key;

for (key in doctypes) {
    shortcodes[cleanDoctype(key)] = doctypes[key];
}

/**
 * Get a doctype from a name.
 *
 * @param {number|string} name - Doctype name.
 * @return {string?} - Doctype.
 */
function getDoctype(name) {
    var key = cleanDoctype(name);
    return shortcodes[key] || shortcodes['h' + key] || null;
}

/*
 * Ensure proper non-versioned types work:
 */

shortcodes.h = shortcodes.h5;
shortcodes.s = shortcodes['s1.1f'];
shortcodes.m = shortcodes.m2;
shortcodes.x = shortcodes['x1.1'];

/*
 * Expose.
 */

getDoctype.all = doctypes;

module.exports = getDoctype;


doctype.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.doctype = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
module.exports={
  "HTML 5": "html",
  "HTML 4.01 Strict": "HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\"",
  "HTML 4.01 Transitional": "HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\"",
  "HTML 4.01 Frameset": "HTML PUBLIC \"-//W3C//DTD HTML 4.01 Frameset//EN\" \"http://www.w3.org/TR/html4/frameset.dtd\"",
  "HTML 3.2": "HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\"",
  "HTML 3.2": "HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\"",
  "HTML 2.0": "HTML PUBLIC \"-//IETF//DTD HTML//EN\"",
  "XHTML 1.0 Strict": "html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\"",
  "XHTML 1.0 Transitional": "html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\"",
  "XHTML 1.0 Frameset": "html PUBLIC \"-//W3C//DTD XHTML 1.0 Frameset//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-frameset.dtd\"",
  "XHTML 1.1": "html PUBLIC \"-//W3C//DTD XHTML 1.1//EN\" \"http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd\"",
  "XHTML Basic 1.1": "html PUBLIC \"-//W3C//DTD XHTML Basic 1.1//EN\" \"http://www.w3.org/TR/xhtml-basic/xhtml-basic11.dtd\"",
  "XHTML Basic 1.0": "html PUBLIC \"-//W3C//DTD XHTML Basic 1.0//EN\" \"http://www.w3.org/TR/xhtml-basic/xhtml-basic10.dtd\"",
  "MathML 2.0": "math PUBLIC \"-//W3C//DTD MathML 2.0//EN\" \"http://www.w3.org/Math/DTD/mathml2/mathml2.dtd\"",
  "MathML 1.01": "math SYSTEM \"http://www.w3.org/Math/DTD/mathml1/mathml.dtd\"",
  "SVG 1.0": "svg PUBLIC \"-//W3C//DTD SVG 1.0//EN\" \"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd\"",
  "SVG 1.1 Full": "svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\"",
  "SVG 1.1 Basic": "svg PUBLIC \"-//W3C//DTD SVG 1.1 Basic//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-basic.dtd\"",
  "SVG 1.1 Tiny": "svg PUBLIC \"-//W3C//DTD SVG 1.1 Tiny//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11-tiny.dtd\""
}

},{}],2:[function(require,module,exports){
/**
 * @author Titus Wormer
 * @copyright 2015 Titus Wormer
 * @license MIT
 * @module doctype
 * @fileoverview Information for HTML doctypes.
 */

'use strict';

/* eslint-env commonjs */

/*
 * Dependencies.
 */

var doctypes = require('./index.json');

/** Get the first character of the second parameter. */
function character(_, value) {
    return value.charAt(0);
}

/** Clean and simplify a doctype name. */
function cleanDoctype(name) {
    return String(name)
        .toLowerCase()
        .replace(/([a-z]+|\d+)/, ' $1 ')
        .replace(/\.0+(?!\d)/, '')
        .replace(/\.01/, '')
        .replace(/\.2/, '')
        .replace(/\b(xhtml|html|mathml|svg|full|basic|tiny|strict|frameset|basic)\b/g, character)
        .replace(/ t(?:ransitional)?/, '')
        .replace(/\s+/g, '');
}

/*
 * Gather all doctypes as shortcodes.
 */

var shortcodes = {};
var key;

for (key in doctypes) {
    shortcodes[cleanDoctype(key)] = doctypes[key];
}

/**
 * Get a doctype from a name.
 *
 * @param {number|string} name - Doctype name.
 * @return {string?} - Doctype.
 */
function getDoctype(name) {
    var key = cleanDoctype(name);
    return shortcodes[key] || shortcodes['h' + key] || null;
}

/*
 * Ensure proper non-versioned types work:
 */

shortcodes.h = shortcodes.h5;
shortcodes.s = shortcodes['s1.1f'];
shortcodes.m = shortcodes.m2;
shortcodes.x = shortcodes['x1.1'];

/*
 * Expose.
 */

getDoctype.all = doctypes;

module.exports = getDoctype;

},{"./index.json":1}]},{},[2])(2)
});